replica/database.cc

```diff
@@ -2833,29 +2799,57 @@ future<> database::drain() {
     b.cancel();
 }
 
+void database::tables_metadata::add_table_helper(database& db, keyspace& ks, table& cf, schema_ptr s) {
+    // A table needs to be added atomically.
+    auto id = s->id();
+    try {
+        ks.add_or_update_column_family(s);
+        s->registry_entry()->set_table(cf.weak_from_this());
+        _column_families.emplace(id, s->table().shared_from_this());
+        _ks_cf_to_uuid.emplace(std::make_pair(s->ks_name(), s->cf_name()), id);
+        if (s->is_view()) {
+            db.find_column_family(s->view_info()->base_id()).add_or_update_view(view_ptr(s));
+        }
+    } catch (...) {
+        // Wrap in noexcept lambda to shutdown on failure.
+        auto revert_changes = [&] () noexcept {
+            remove_table_helper(db, ks, cf, s);
+        };
+        revert_changes();
```

Comment:
Is this revert_changes guaranteed to work regardless of where the exception was thrown?

A pattern that works well is to use defer():

```c++
do_first_thing();
auto undo1 = defer([&] { undo_first_thing(); });
do_second_thing();
auto undo2 = defer([&] { undo_second_thing(); });
...
undo1.cancel();
undo2.cancel();
...
```

Reply:
It will, but I guess defers would be safer in case it's be modified, changing